text/2022-02-14-link-application-metrics-with-chaos-experiments.md

```diff
@@ -0,0 +1,212 @@
+# Link Application Metrics with Chaos Experiments
+
+## Summary
+
+Chaos Dashboard could link Chaos Experiment with Grafana Panels, and display the
+application metrics relates to the chaos experiment(also the schedule, and
+workflow).
+
+## Motivation
+
+<!-- Why are we doing this? What use cases does it support? What is the expected
+outcome? -->
+
+It's a feature drawback that we could not measure the steady-state of the
+application. It's one of the important steps in Chaos Engineering Principles.
+
+Although it's very hard to make assertions on applications's steady-state
+automatically, we could make some improvements to help people/user to make
+assertions on the steady-state.
+
+## Detailed design
+
+<!-- This is the bulk of the RFC. Explain the design in enough detail that:
+
+- It is reasonably clear how the feature would be implemented.
+- Corner cases are dissected by example.
+- How the feature is used. -->
+
+### Overview
+
+This feature is all about Chaos Dashboard, would not affect anything with other
+components and CRD.
+
+We would use Grafana as the source of application metrics, and use Grafana Panel
```

Comment:
This seems weird. If the goal is to correlate chaos experiments with user applications, then why we don't just correlate with the metrics directly? Grafana itself is the UI/display layer.

Reply:
I will try to answer all of the following questions together, but my understanding may not be as accurate as @STRRL and @cwen0, feel free to @ them.

Overall, this is more of an exploration. The goal is to bring users an integrated interface that includes experiments, workflows, and panels rendered from data obtained through the grafana API.

To achieve this, we want to be able to fetch the grafana panel data through the grafana API and then customize the rendering in the chaos dashboard. We will take the panel as the smallest unit and store it as a resource so that any chaos can associate the panel and combine their active time with the original panel data.

Users can simply click `Share` this panel, paste the link into the chaos dashboard's interface, and associate anything Chaos Mesh related with the panel's data.

It's an all-in-one experience where users no longer need to install plugins, configure the API address, and complicate their panels.

But that doesn't mean that we're going to abandon grafana datasource, we want to make it more of a community-driven project, and any good ideas and suggestions we have will be added to the datasource plugin if any.

Reply:
But I still want to make the point that getting close to the grafana ecosystem by way of plugins is extremely difficult, and as someone who has had hands-on experience writing plugins is painful (there is no extensive documentation, and I had to try to write through some plugin source code). I think this will make it fundamentally difficult to attract more people to plugin development.